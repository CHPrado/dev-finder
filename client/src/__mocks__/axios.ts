const axios = {
  get: jest.fn(() => Promise.resolve({})),
  post: jest.fn(() => Promise.resolve({})),
  create: () => axios,
  defaults: {
    adapter: {},
  },
};

export default axios;
