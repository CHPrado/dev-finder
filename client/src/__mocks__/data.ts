import { createMemoryHistory, createLocation } from "history";

import { DevProps, DevRouteProps } from "../interfaces";

const devData: DevProps = {
  username: "CHPrado",
  name: "Christian Prado",
  bio: "My bio",
  avatar_url: "avatar_url",
  location: "Lins - SP",
  followers: 1,
  following: 2,
  public_repos: 3,
  stars: 4,
  watchers: 5,
  forks: 6,
  languages: [
    {
      name: "JavaScript",
      count: 1,
      percent: "49.9",
    },
    {
      name: "Ruby",
      count: 1,
      percent: "51.1",
    },
  ],
  notes: "Notes about me",
};

const devRouteData: DevRouteProps = {
  history: createMemoryHistory(),
  location: createLocation({ state: { dev: devData } }),
  match: {
    isExact: false,
    path: "/dev",
    url: "/dev",
    params: {},
  },
};

export { devData, devRouteData };
