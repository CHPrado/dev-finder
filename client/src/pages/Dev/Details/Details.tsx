import React, { FC } from "react";
import { Link } from "react-router-dom";

import { Box, Typography } from "@material-ui/core";
import {
  RepoIcon,
  StarIcon,
  PeopleIcon,
  GitForkIcon,
  EyeIcon,
} from "@primer/octicons-react";

import { DevProps } from "../../../interfaces";
import useStyles from "./styles";

interface Props {
  dev: DevProps;
}

const Details: FC<Props> = ({ dev }) => {
  const classes = useStyles();

  return (
    <Box className={classes.root}>
      <Box width="100%">
        <Typography className={classes.value}>{dev.followers}</Typography>
        <Typography className={classes.value}>{dev.following}</Typography>
        <Typography className={classes.value}>{dev.public_repos}</Typography>
        <Typography className={classes.value}>{dev.stars}</Typography>
        <Typography className={classes.value}>{dev.watchers}</Typography>
        <Typography className={classes.value}>{dev.forks}</Typography>
      </Box>

      <Box display="flex" flexDirection="column" width="100%">
        <Link
          className={classes.link}
          to={(location) => ({
            ...location,
            pathname: `http://www.github.com/${dev.username}?tab=followers`,
          })}
          target="_blank"
        >
          <Typography className={classes.text}>
            <PeopleIcon size={16} className={classes.icon} />
            followers
          </Typography>
        </Link>
        <Link
          className={classes.link}
          to={(location) => ({
            ...location,
            pathname: `http://www.github.com/${dev.username}?tab=following`,
          })}
          target="_blank"
        >
          <Typography className={classes.text}>
            <PeopleIcon size={16} className={classes.icon} />
            following
          </Typography>
        </Link>
        <Link
          className={classes.link}
          to={(location) => ({
            ...location,
            pathname: `http://www.github.com/${dev.username}?tab=repositories`,
          })}
          target="_blank"
        >
          <Typography className={classes.text}>
            <RepoIcon size={16} className={classes.icon} />
            public repositories
          </Typography>
        </Link>
        <Link
          className={classes.link}
          to={(location) => ({
            ...location,
            pathname: `http://www.github.com/${dev.username}?tab=stars`,
          })}
          target="_blank"
        >
          <Typography className={classes.text}>
            <StarIcon size={16} className={classes.icon} />
            stars
          </Typography>
        </Link>
        <Typography className={classes.text}>
          <EyeIcon size={16} className={classes.icon} />
          watchers
        </Typography>
        <Typography className={classes.text}>
          <GitForkIcon size={16} className={classes.icon} />
          forks
        </Typography>
      </Box>
    </Box>
  );
};

export default Details;
