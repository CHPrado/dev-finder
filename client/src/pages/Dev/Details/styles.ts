import { Theme, makeStyles } from "@material-ui/core/styles";

export default makeStyles((theme: Theme) => ({
  root: {
    display: "flex",
    marginRight: theme.spacing(2),
    [theme.breakpoints.down("xs")]: {
      margin: 0,
    },
  },
  value: {
    fontWeight: theme.typography.fontWeightBold,
    color: theme.palette.text.secondary,
    margin: theme.spacing(0, 1),
    textAlign: "right",
    display: "flex",
    alignItems: "center",
    justifyContent: "flex-end",
  },
  link: {
    textDecoration: "none",
    "& :hover": {
      color: theme.palette.primary.main,
      textDecoration: "underline",
    },
  },
  text: {
    color: theme.palette.text.secondary,
    margin: theme.spacing(0, 1),
    whiteSpace: "nowrap",
  },
  icon: {
    marginRight: theme.spacing(2),
  },
}));
