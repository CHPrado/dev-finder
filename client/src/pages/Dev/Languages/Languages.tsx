import React, { FC } from "react";

import { Box, Typography } from "@material-ui/core";

import { DevProps } from "../../../interfaces";
import useStyles from "./styles";

interface Props {
  languages: DevProps["languages"];
}

const Languages: FC<Props> = ({ languages }) => {
  const classes = useStyles();

  return (
    <Box mt={{ xs: 4, sm: 0 }}>
      {languages.map((language) => (
        <Box key={language.name} className={classes.languageBox}>
          <Box width="100%">
            <Typography className={classes.language}>
              {language.name}
            </Typography>
          </Box>
          <Box width="100%">
            <Typography className={classes.percent}>
              {language.percent}%
            </Typography>
          </Box>
        </Box>
      ))}
    </Box>
  );
};

export default Languages;
