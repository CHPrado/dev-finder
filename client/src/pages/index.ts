export { default as Home } from "./Home";
export { default as Dev } from "./Dev";
