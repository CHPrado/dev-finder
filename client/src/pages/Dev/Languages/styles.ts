import { Theme, makeStyles } from "@material-ui/core/styles";

export default makeStyles((theme: Theme) => ({
  languageBox: {
    display: "flex",
    justifyContent: "space-between",
    [theme.breakpoints.down("xs")]: {
      justifyContent: "center",
    },
  },
  language: {
    color: theme.palette.text.secondary,
    margin: theme.spacing(0, 1),
    whiteSpace: "nowrap",
    [theme.breakpoints.down("xs")]: {
      textAlign: "right",
    },
  },
  percent: {
    fontWeight: theme.typography.fontWeightBold,
    color: theme.palette.text.secondary,
    margin: theme.spacing(0, 1),
    textAlign: "right",
    [theme.breakpoints.down("xs")]: {
      textAlign: "left",
    },
  },
}));
