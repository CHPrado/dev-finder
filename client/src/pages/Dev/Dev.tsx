import React, { FC, useState } from "react";
import { Link } from "react-router-dom";

import {
  Box,
  Card,
  CircularProgress,
  IconButton,
  Tooltip,
  Typography,
} from "@material-ui/core";
import { CachedOutlined } from "@material-ui/icons";
import { LocationIcon } from "@primer/octicons-react";

import { Header, Snackbar } from "../../components";
import { SnackbarProps, DevRouteProps } from "../../interfaces";
import { githubApi } from "../../services";
import { Details } from "./Details";
import { Languages } from "./Languages";
import { Notes } from "./Notes";
import useStyles from "./styles";

const Dev: FC<DevRouteProps> = (props) => {
  const classes = useStyles();
  const [dev, setDev] = useState(props.location.state.dev);
  const [isLoading, setIsLoading] = useState(false);
  const [snackbar, setSnackbar] = useState<SnackbarProps>({});

  const handleRefresh = () => {
    setIsLoading(true);
    githubApi
      .getDev(dev.username)
      .then((resp) => {
        setDev(resp);
      })
      .catch(() => {
        setSnackbar({
          open: true,
          message:
            "There was an error trying to refresh the data. Please, try again.",
          type: "error",
          onClose: () => setSnackbar({ open: false }),
        });
      })
      .finally(() => setIsLoading(false));
  };

  return (
    <Box className={classes.root}>
      <Header />

      {isLoading ? (
        <Box className={classes.loadingBox}>
          <CircularProgress size="100px" />
        </Box>
      ) : (
        <Box className={classes.container}>
          <Card variant="outlined" className={classes.card}>
            <Box className={classes.titleBox}>
              <Link
                className={classes.link}
                to={(location) => ({
                  ...location,
                  pathname: `http://www.github.com/${dev.username}`,
                })}
                target="_blank"
              >
                <Typography className={classes.name}>
                  {`${dev.username}'s profile`}
                </Typography>
              </Link>

              <Box display="flex">
                <Tooltip title="Refresh">
                  <IconButton onClick={handleRefresh}>
                    <CachedOutlined />
                  </IconButton>
                </Tooltip>
              </Box>
            </Box>

            <Box className={classes.avatarBox}>
              <img
                src={dev.avatar_url}
                alt="avatar"
                className={classes.avatar}
              />

              <Box className={classes.infoBox}>
                <Typography className={classes.info}>{dev.name}</Typography>
                {dev.location && (
                  <Typography className={classes.info}>
                    <LocationIcon size={18} className={classes.icon} />
                    {dev.location}
                  </Typography>
                )}
                <Typography color="textSecondary" align="justify">
                  {dev.bio}
                </Typography>
              </Box>
            </Box>

            <Box className={classes.detailsBox}>
              <Details dev={dev} />
              <Languages languages={dev.languages} />
            </Box>

            <Notes
              dev={dev}
              setIsLoading={setIsLoading}
              setSnackbar={setSnackbar}
              setDev={setDev}
            />
          </Card>
        </Box>
      )}

      <Snackbar {...snackbar} />
    </Box>
  );
};

export default Dev;
