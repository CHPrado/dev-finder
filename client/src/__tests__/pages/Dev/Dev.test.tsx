import React from "react";
import { BrowserRouter } from "react-router-dom";

import { IconButton, TextField } from "@material-ui/core";
import { act } from "@testing-library/react";
import axios from "axios";
import { mount, shallow } from "enzyme";
import toJson from "enzyme-to-json";

import { devData, devRouteData } from "../../../__mocks__/data";
import { Dev } from "../../../pages";
import { Details } from "../../../pages/Dev/Details";
import { Languages } from "../../../pages/Dev/Languages";

jest.mock("axios");

const devWrapper = mount(
  <BrowserRouter>
    <Dev {...devRouteData} />
  </BrowserRouter>
);

describe("rendering component", () => {
  it("renders Dev component without crashing", () => {
    const headerTitle = devWrapper.find("h4").render().text();
    const content = devWrapper.render().text();

    expect(headerTitle).toEqual("lookingforDev");
    expect(content).toContain(devData.username);
    expect(content).toContain(devData.name);
    expect(content).toContain(devData.location);
    expect(content).toContain(devData.bio);
    expect(content).toContain(devData.notes);
    expect(content).toContain(devData.languages[0].name);
    expect(content).toContain(devData.languages[1].name);
  });

  it("renders Details component without crashing", () => {
    const details = devWrapper.find(Details).at(0);

    expect(details.props().dev).toEqual(devData);
  });

  it("renders Languages component without crashing", () => {
    const languages = devWrapper.find(Languages).at(0);

    expect(languages.props().languages).toEqual(devData.languages);

    expect(languages.contains(devData.languages[0].name)).toEqual(true);
    expect(languages.contains(`${devData.languages[0].percent}`)).toEqual(true);
    expect(languages.contains(devData.languages[1].name)).toEqual(true);
    expect(languages.contains(`${devData.languages[1].percent}`)).toEqual(true);
  });

  it("renders Notes component without crashing", () => {
    const notes = devWrapper.find(TextField).at(0).props().value;

    expect(notes).toEqual("Notes about me");
  });
});

describe("passing props", () => {
  it("dev props data equals to dev data", () => {
    const dev = devWrapper.find(Dev).at(0);

    expect(dev.props().location.state.dev).toEqual(
      devRouteData.location.state.dev
    );
  });
});

describe("simulating events", () => {
  it("simulates refresh button click", async () => {
    const refreshButton = devWrapper.find(IconButton).at(0);
    const getSpy = jest.spyOn(axios, "get");

    await act(async () => {
      refreshButton.simulate("click");
    });

    expect(getSpy).toBeCalled();
  });
});

describe("snapshots", () => {
  it("Dev snapshots", () => {
    const tree = shallow(<Dev {...devRouteData} />);

    expect(toJson(tree)).toMatchSnapshot();
  });
});
