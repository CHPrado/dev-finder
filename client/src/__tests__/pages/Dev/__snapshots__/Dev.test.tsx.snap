// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshots Dev snapshots 1`] = `
<Styled(MuiBox)
  className="makeStyles-root-1"
>
  <Header />
  <Styled(MuiBox)
    className="makeStyles-container-3"
  >
    <WithStyles(ForwardRef(Card))
      className="makeStyles-card-4"
      variant="outlined"
    >
      <Styled(MuiBox)
        className="makeStyles-titleBox-5"
      >
        <Link
          className="makeStyles-link-13"
          target="_blank"
          to={[Function]}
        >
          <WithStyles(ForwardRef(Typography))
            className="makeStyles-name-6"
          >
            CHPrado's profile
          </WithStyles(ForwardRef(Typography))>
        </Link>
        <Styled(MuiBox)
          display="flex"
        >
          <WithStyles(ForwardRef(Tooltip))
            title="Refresh"
          >
            <WithStyles(ForwardRef(IconButton))
              onClick={[Function]}
            >
              <Memo([object Object]) />
            </WithStyles(ForwardRef(IconButton))>
          </WithStyles(ForwardRef(Tooltip))>
        </Styled(MuiBox)>
      </Styled(MuiBox)>
      <Styled(MuiBox)
        className="makeStyles-avatarBox-7"
      >
        <img
          alt="avatar"
          className="makeStyles-avatar-8"
          src="avatar_url"
        />
        <Styled(MuiBox)
          className="makeStyles-infoBox-9"
        >
          <WithStyles(ForwardRef(Typography))
            className="makeStyles-info-10"
          >
            Christian Prado
          </WithStyles(ForwardRef(Typography))>
          <WithStyles(ForwardRef(Typography))
            className="makeStyles-info-10"
          >
            <LocationIcon
              className="makeStyles-icon-12"
              size={18}
              verticalAlign="text-bottom"
            />
            Lins - SP
          </WithStyles(ForwardRef(Typography))>
          <WithStyles(ForwardRef(Typography))
            align="justify"
            color="textSecondary"
          >
            My bio
          </WithStyles(ForwardRef(Typography))>
        </Styled(MuiBox)>
      </Styled(MuiBox)>
      <Styled(MuiBox)
        className="makeStyles-detailsBox-11"
      >
        <Details
          dev={
            Object {
              "avatar_url": "avatar_url",
              "bio": "My bio",
              "followers": 1,
              "following": 2,
              "forks": 6,
              "languages": Array [
                Object {
                  "count": 1,
                  "name": "JavaScript",
                  "percent": "49.9",
                },
                Object {
                  "count": 1,
                  "name": "Ruby",
                  "percent": "51.1",
                },
              ],
              "location": "Lins - SP",
              "name": "Christian Prado",
              "notes": "Notes about me",
              "public_repos": 3,
              "stars": 4,
              "username": "CHPrado",
              "watchers": 5,
            }
          }
        />
        <Languages
          languages={
            Array [
              Object {
                "count": 1,
                "name": "JavaScript",
                "percent": "49.9",
              },
              Object {
                "count": 1,
                "name": "Ruby",
                "percent": "51.1",
              },
            ]
          }
        />
      </Styled(MuiBox)>
      <Notes
        dev={
          Object {
            "avatar_url": "avatar_url",
            "bio": "My bio",
            "followers": 1,
            "following": 2,
            "forks": 6,
            "languages": Array [
              Object {
                "count": 1,
                "name": "JavaScript",
                "percent": "49.9",
              },
              Object {
                "count": 1,
                "name": "Ruby",
                "percent": "51.1",
              },
            ],
            "location": "Lins - SP",
            "name": "Christian Prado",
            "notes": "Notes about me",
            "public_repos": 3,
            "stars": 4,
            "username": "CHPrado",
            "watchers": 5,
          }
        }
        setDev={[Function]}
        setIsLoading={[Function]}
        setSnackbar={[Function]}
      />
    </WithStyles(ForwardRef(Card))>
  </Styled(MuiBox)>
  <Snackbar />
</Styled(MuiBox)>
`;
